### PR TITLE
Layer size specified in the manifest must match actual blob size

### PR DESCRIPTION
**Pull request by legionus**

### CODE REVIEW

File: pkg/dockerregistry/server/manifestschema1handler.go [hunk 1]
```diff
@@ -90,9 +90,11 @@ func (h *manifestSchema1Handler) Verify(ctx context.Context, skipDependencyVerif
 			})
 	}
 
-	if len(h.manifest.History) != len(h.manifest.FSLayers) {
-		errs = append(errs, fmt.Errorf("mismatched history and fslayer cardinality %d != %d",
-			len(h.manifest.History), len(h.manifest.FSLayers)))
+	manifestReferences := h.manifest.References()
+
+	if len(h.manifest.History) != len(manifestReferences) {
+		errs = append(errs, fmt.Errorf("mismatched history and manifests references %d != %d",
```
Comment (miminar) [inline pkg/dockerregistry/server/manifestschema1handler.go]: manifest reference cardinality

File: pkg/dockerregistry/server/manifestschema1handler.go [hunk 2]
```diff
@@ -125,11 +129,37 @@ func (h *manifestSchema1Handler) Verify(ctx context.Context, skipDependencyVerif
...
+		Size int64 `json:"size,omitempty"`
+	}{}
+
+	for i, obj := range h.manifest.History {
+		v1Compatibility.Size = 0
+		if err := json.Unmarshal([]byte(obj.V1Compatibility), &v1Compatibility); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if !isEmptyDigest(manifestReferences[i].Digest) && realSizes[i] != v1Compatibility.Size {
```
Comment (miminar) [inline pkg/dockerregistry/server/manifestschema1handler.go]: I don't think this verification can be applied to schema 1 manifests.
I'm almost sure that blob size stored in the registry does not match the size in manifest history. Docker computes (or used to compute) this size from the extracted data.

File: pkg/dockerregistry/server/manifestservice.go [hunk 3]
```diff
@@ -21,6 +21,20 @@ import (
 	quotautil "github.com/openshift/origin/pkg/quota/util"
 )
 
+// ErrManifestBlobBadSize returned when a size of blob specified in the manifest
```
Comment (miminar) [inline pkg/dockerregistry/server/manifestservice.go]: is returned

File: pkg/dockerregistry/server/manifestservice.go [hunk 4]
```diff
@@ -21,6 +21,20 @@ import (
...
+// does not match the actual size. The docker/distribution does not check this
+// and therefore does not have a ready-made error for this.
+type ErrManifestBlobBadSize struct {
```
Comment (miminar) [inline pkg/dockerregistry/server/manifestservice.go]: Is it worth declaring and returning when turned into 500 error?

File: pkg/dockerregistry/server/manifestschema2handler.go [hunk 5]
```diff
@@ -85,21 +85,28 @@ func (h *manifestSchema2Handler) Verify(ctx context.Context, skipDependencyVerif
...
 	if err != nil {
 		if err != distribution.ErrBlobUnknown {
 			errs = append(errs, err)
 		}
 
 		// On error here, we always append unknown blob errors.
 		errs = append(errs, distribution.ErrManifestBlobUnknown{Digest: target.Digest})
+	} else {
+		if target.Size != desc.Size {
+			errs = append(errs, ErrManifestBlobBadSize{
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/dockerregistry/server/manifestschema2handler.go]: > Or maybe wrap this in ErrManifestVerification.

@miminar Em... but `errs` is already `distribution.ErrManifestVerification`

File: pkg/dockerregistry/server/manifestschema1handler.go (hunk 1 above)
Comment (legionus) [reply pkg/dockerregistry/server/manifestschema1handler.go]: @miminar I'm not sure I understand.

File: pkg/dockerregistry/server/manifestschema1handler.go (hunk 2 above)
Comment (legionus) [reply pkg/dockerregistry/server/manifestschema1handler.go]: @miminar It must match or we can not use this field anywhere at all.

File: pkg/dockerregistry/server/manifestservice.go (hunk 4 above)
Comment (legionus) [reply pkg/dockerregistry/server/manifestservice.go]: @miminar This error will be part of `ErrManifestVerification`.

Comment (legionus) [reply pkg/dockerregistry/server/manifestservice.go]: @miminar It's ok for upstream to put unknown errors:

https://github.com/openshift/origin/blob/master/vendor/github.com/docker/distribution/registry/storage/signedmanifesthandler.go#L106

File: pkg/dockerregistry/server/manifestschema2handler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmage) [reply pkg/dockerregistry/server/manifestschema2handler.go]: @legionus https://github.com/docker/distribution/blob/7484e51bf6af0d3b1a849644cdaced3cfcf13617/registry/handlers/manifests.go#L312-L328

[the remark this answers is not in the document]
Comment (dmage) [reply pkg/dockerregistry/server/manifestschema2handler.go]: you can wrap it with v2.ErrorCodeManifestInvalid.WithDetail(err)

File: pkg/dockerregistry/server/manifestservice.go (hunk 4 above)
Comment (dmage) [reply pkg/dockerregistry/server/manifestservice.go]: @legionus for user it'll be "unknown error"

File: pkg/dockerregistry/server/manifestschema1handler.go (hunk 2 above)
Comment (dmage) [reply pkg/dockerregistry/server/manifestschema1handler.go]: @legionus why do you need isEmptyDigest here?

Comment (dmage) [inline pkg/dockerregistry/server/manifestschema1handler.go]: @legionus why do you need isEmptyDigest here?

File: pkg/dockerregistry/server/manifestschema2handler.go [hunk 6]
```diff
@@ -85,21 +85,28 @@ func (h *manifestSchema2Handler) Verify(ctx context.Context, skipDependencyVerif
...
+				Digest:       target.Digest,
+				ActualSize:   desc.Size,
+				ManifestSize: target.Size,
+			})
+		}
 	}
 
 	for _, fsLayer := range h.manifest.References() {
-		var err error
```
Comment (dmage) [inline pkg/dockerregistry/server/manifestschema2handler.go]: Return this line or reset the value of err

File: pkg/dockerregistry/server/manifestschema2handler.go [hunk 7]
```diff
@@ -118,6 +125,14 @@ func (h *manifestSchema2Handler) Verify(ctx context.Context, skipDependencyVerif
 
 			// On error here, we always append unknown blob errors.
 			errs = append(errs, distribution.ErrManifestBlobUnknown{Digest: fsLayer.Digest})
+			continue
+		}
+		if fsLayer.Size != desc.Size {
```
Comment (dmage) [inline pkg/dockerregistry/server/manifestschema2handler.go]: You shouldn't check it when `fsLayer.MediaType == schema2.MediaTypeForeignLayer`.

File: pkg/dockerregistry/server/manifestschema1handler.go (hunk 2 above)
Comment (miminar) [reply pkg/dockerregistry/server/manifestschema1handler.go]: For schema 2 it's relatively safe. For schema 1:

<details>

```
$ docker run -it --rm --net=host docker:1.8 -H tcp://0.0.0.0:2375 version
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 18:01:15 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 18:01:15 UTC 2015
 OS/Arch:      linux/amd64

$ docker -H tcp://0.0.0.0:2375 push localhost:5002/fedora23
The push refers to a repository [localhost:5002/fedora23] (len: 1)
d295b2374c0b: Image successfully pushed
904845c49256: Image successfully pushed
8289b4c23708: Image successfully pushed
latest: digest: sha256:fca2e2d21f87264e171ea7dcb58e6f5ec985150fb503bd70105933176caf8b08 size: 4020
```

</details>

Manifest: 

<details>

```
$ cat blobs/sha256/fc/fca2e2d21f87264e171ea7dcb58e6f5ec985150fb503bd70105933176caf8b08/data
{
   "schemaVersion": 1,
   "name": "fedora23",
   "tag": "latest",
   "architecture": "amd64",
   "fsLayers": [
      {
         "blobSum": "sha256:251d5f2b1db1b45487cf714b7d0be22ea552e5ad03bc7f995cfe3f4900efa7e0"
      },
      {
         "blobSum": "sha256:14707640923fd877a8f157764c21e9bb0d003560c9928a3d40ad2454bdf8af1e"
      },
      {
         "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
      }
   ],
   "history": [
      {
         "v1Compatibility": "{\"id\":\"d295b2374c0bc1f3cf377363fb0e47f2bc1b346451f43c6ccfb9088af9acb080\",\"parent\":\"
388cb781fe27aa05c674cd40eeab5219839b4988998c8eafedeb917ac3c9c810\",\"created\":\"2017-10-13T12:08:16.68090848Z\",\"cont
ainer\":\"292f96cfefecea84a53b0ace771b89c67f1f954e057849e09f0aac305d57d8e9\",\"container_config\":{\"Hostname\":\"1b71d
b504850\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"Tty\"
:false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:
/bin\"],\"Cmd\":[\"/bin/sh\",\"-c\",\"#(nop) COPY file:3ae3457f7dbd1a373e995cb54b2a7e1fb77f73c2cd4fce419ae62cd8792c0992
 in /data\"],\"Image\":\"904845c492565ba38ef083ad60136f1ae3d91343aad64a150fb82754e00073e2\",\"Volumes\":null,\"WorkingD
ir\":\"\",\"Entrypoint\":null,\"OnBuild\":[],\"Labels\":{}},\"docker_version\":\"1.8.3\",\"config\":{\"Hostname\":\"1b7
1db504850\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"Tty
\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbi
n:/bin\"],\"Cmd\":null,\"Image\":\"904845c492565ba38ef083ad60136f1ae3d91343aad64a150fb82754e00073e2\",\"Volumes\":null,
\"WorkingDir\":\"\",\"Entrypoint\":null,\"OnBuild\":[],\"Labels\":{}},\"architecture\":\"amd64\",\"os\":\"linux\",\"Siz
e\":51200}"
      },
      {
         "v1Compatibility": "{\"architecture\":\"amd64\",\"author\":\"[Adam Miller \\u003cmaxamillion@fedoraproject.org
\\u003e] [Patrick Uiterwijk \\u003cpatrick@puiterwijk.org\\u003e]\",\"config\":{\"Hostname\":\"1b71db504850\",\"Domainn
ame\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"Tty\":false,\"OpenStdin\
":false,\"StdinOnce\":false,\"Env\":[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"],\"Cmd\":nul
l,\"Image\":\"sha256:19b1996112de75e361b549a487203ed53cb2cbaa168f42fb700491b598712404\",\"Volumes\":null,\"WorkingDir\"
:\"\",\"Entrypoint\":null,\"OnBuild\":null,\"Labels\":{}},\"container\":\"1b71db504850c430ffd8b6cf43027b77cfebd70e39d40
d0afcc34f596dcfbd58\",\"container_config\":{\"Hostname\":\"1b71db504850\",\"Domainname\":\"\",\"User\":\"\",\"AttachStd
in\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":
[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"],\"Cmd\":[\"/bin/sh\",\"-c\",\"#(nop) ADD file:6
de4fbfd61cdc68ac14c9e5a50007f21d1535a378383c38abac7de569c0a90d3 in / \"],\"Image\":\"sha256:19b1996112de75e361b549a4872
03ed53cb2cbaa168f42fb700491b598712404\",\"Volumes\":null,\"WorkingDir\":\"\",\"Entrypoint\":null,\"OnBuild\":null,\"Lab
els\":{}},\"created\":\"2016-08-30T18:21:35.691689327Z\",\"docker_version\":\"1.12.1\",\"id\":\"388cb781fe27aa05c674cd4
0eeab5219839b4988998c8eafedeb917ac3c9c810\",\"os\":\"linux\",\"parent\":\"3690474eb5b4b26fdfbd89c6e159e8cc376ca76ef4803
2a30fa6aafd56337880\"}"
      },
      {
         "v1Compatibility": "{\"id\":\"3690474eb5b4b26fdfbd89c6e159e8cc376ca76ef48032a30fa6aafd56337880\",\"created\":\
"2016-08-30T18:21:14.673847011Z\",\"container_config\":{\"Cmd\":[\"/bin/sh -c #(nop)  MAINTAINER [Adam Miller \\u003cma
xamillion@fedoraproject.org\\u003e] [Patrick Uiterwijk \\u003cpatrick@puiterwijk.org\\u003e]\"]},\"author\":\"[Adam Mil
ler \\u003cmaxamillion@fedoraproject.org\\u003e] [Patrick Uiterwijk \\u003cpatrick@puiterwijk.org\\u003e]\",\"throwaway
\":true}"
      }
   ]
```

</details>

Note the `\"Size\":51200` of the last layer (the one at index 0). This is the size of the uncompressed data I added to the fedora:23 image.

Now see the actual size of the blob:

```
$ ls -l blobs/sha256/25/251d5f2b1db1b45487cf714b7d0be22ea552e5ad03bc7f995cfe3f4900efa7e0/data
-rw-r--r-- 1 root root 51412 13. Okt 14:10 blobs/sha256/25/251d5f2b1db1b45487cf714b7d0be22ea552e5ad03bc7f995cfe3f4900efa7e0/data
```

For schema 1, sizes do not match.

**Comment by legionus** [comment]
We have a better idea.

**Pull request closed without merging**